DIPs/DIP1xxx-rn.md

```diff
@@ -0,0 +1,927 @@
+to the `opAssign` one:
+
+```d
+struct A
+{
+    int a;
+    this(int a)
+    {
+        this.a = a;
+    }
+    @implicit this(ref A rhs)
+    {
+        this.a = rhs.a;
+    }
+    void opAssign(S rhs)
+    {
+        this.a = rhs.a;
+    }
+}
+
+void main()
+{
+    A a = A(2);
+    A b = a;      // calls copy constructor;
+    b = a;        // calls opAssign;
+}
+```
+
+In order to avoid the code duplication resulting from such situations, ideally, the user could
+define a single method that deals with both copy construction and normal copying. This is possible
+in certain situations which will be discussed below:
+
+1. If the struct contains any `const`/`immutable` fields, it is impossible to use the copy constructor
+for `opAssign`, because the copy constructor might initialize them. Even if the copy constructor doesn't
+modify the `const`/`immutable` fields the compiler has to analyze the function body to know that, which
+is problematic in situations when the body is missing. In conclusion, `opAssign` can be generated from the
+copy constructor when the `struct` contains only assignable (mutable) fields.
+
+2. The copy constructor signature is : `@implicit this(ref $q1 S rhs) $q2`, where `q1` and `q2` represent the
+qualifiers that can be applied to the function and the parameter (`const`, `immutable`, `shared`, etc.). The
+problem that arises is: depending on the values of `$q1` and `$q2` what should the signature of `opAssign` be?
+A solution might be to generate for every copy constructor its counterpart `opAssign`: `void opAssign(ref $q1 S rhs) $q2`.
+However, when is a `const`/`immutable` `opAssign` needed? There might be obscure cases when that is useful, but those are
+niche situations where the user must step it and clarify what the desired outcome is and define its own `opAssign`. For
+the sake of simplicity, `opAssign` will be generated solely for copy constructors that have a missing `$q2`.
+
+Taking into account the above mentioned restrictions we can generate a single `opAssign` function
+and leverage the existing `opAssign` generation logic for the [[postblit](https://github.com/dlang/dmd/pull/8505)].
+Because `opAssing` is generated only if there are copy constructors that create a mutable object, the copy
+constructor is going to be called (if possible) when the source is passed as a parameter to the `opAssign`
+function.
+
+## Breaking Changes and Deprecations
+
+1. The parameter of the copy constructor is passed by a mutable reference to the
+source object. This means that a call to the copy constructor may legally modify
+the source object:
+
+```d
+struct A
+{
+    int[] a;
+    @implicit this(ref A another)
+    {
+        another.a[2] = 3;
+    }
+}
+
+void main()
+{
+    A a, b;
+    a = b;    // b.a[2] is modified
+}
+```
+
+A solution to this might be to make the reference `const`, but that would make code like
+`this.a = another.a` inside the copy constructor illegal. Of course, this can be solved by means
```

Comment:
s/would/"could possibly"

depends on the types in A (i.e. if it was a type with no indirections then we all good with a const ref).

Is this in the correct section? This is not a breaking change or a deprecation is it?

Reply:
If you take into account that the postblit does not offer the possibility of modifying the source, it can be seen as a breaking change. I'm sure that some folks might see it as a feature :D :dagger: